src/hotspot/share/opto/type.cpp

```diff
@@ -4036,15 +4036,18 @@ int TypeInstPtr::hash(void) const {
 //------------------------------dump2------------------------------------------
 // Dump oop Type
 #ifndef PRODUCT
-void TypeInstPtr::dump2( Dict &d, uint depth, outputStream *st ) const {
+void TypeInstPtr::dump2( Dict &d, uint depth, outputStream* st ) const {
   // Print the name of the klass.
   klass()->print_name_on(st);
 
   switch( _ptr ) {
   case Constant:
-    // TO DO: Make CI print the hex address of the underlying oop.
     if (WizardMode || Verbose) {
-      const_oop()->print_oop(st);
+      stringStream ss;
+      ss.print(" ");
+      const_oop()->print_oop(&ss);
+      ss.tr_delete('\n');
```

Comment:
`tr_delete` is expensive. Also deleting something in a stream does not fit into a concept of streams.
I see that the content of `ss` is traversed many times.
What about this code:
```
for (const char *str = ss.base(); *str; ) {
  size_t i = 0;
  while (str[i] && str[i] != '\n' ) {
       ++i;
  }
  st->print_raw(str, i);
  str += i;
  while (*str == '\n') {
    ++str;
  }
}
```

Reply:
Another option:
```
class filterStringStream: public stringStream {
private:
  char ch;
public:
  filterStringStream(char ch_to_filter, size_t initial_bufsize = 256) : stringStream(initial_bufsize) {}

  virtual void write(const char* c, size_t len) override {
    const char* e = c + len;
    while (c != e) {
      size_t i = 0;
      while ((c+i) != e && c[i] != ch ) {
        ++i;
      }
      stringStream::write(c, i);
      c += i;
      while (c != e && *ch == ch) {
        ++c;
      }
    }     
  }
};
```

Your code will be:
```
filterStringStream ss('\n');
ss.print(" ");
const_oop->print_oop(&ss);
st->print_raw(ss.base(), ss,size());
```

Reply:
> `tr_delete` is expensive. Also deleting something in a stream does not fit into a concept of streams.
> I see that the content of `ss` is traversed many times.
> What about this code:
> 
> ```
> for (const char *str = ss.base(); *str; ) {
>   size_t i = 0;
>   while (str[i] && str[i] != '\n' ) {
>        ++i;
>   }
>   st->print_raw(str, i);
>   str += i;
>   while (*str == '\n') {
>     ++str;
>   }
> }
> ```

You can put this code in a function like `print_filtering_ch(char, const stringStream&, outputStream*)`

Reply:
hi, @eastig ,
 
Thank you for reviewing this code. 

you are right, I shouldn't modify the contents of a stringStream. I treated it as a buffer instead of stream. 
I took you advice, I moved `tr_delete` logic to StringUtils, which is a toolkit class. 
Could you take a look again?